services/rds/src/main/java/software/amazon/awssdk/services/rds/RdsUtilities.java

```diff
@@ -0,0 +1,139 @@
+                                            .encodedPath("/")
+                                            .putRawQueryParameter("DBUser", request.userName())
+                                            .putRawQueryParameter("Action", "connect")
+                                            .build();
+
+        Instant expirationTime = Instant.now().plus(15, ChronoUnit.MINUTES);
+        Aws4PresignerParams presignRequest = Aws4PresignerParams.builder()
+                                                .expirationTime(expirationTime)
+                                                .awsCredentials(resolveCredentials(request).resolveCredentials())
+                                                .signingName("rds-db")
+                                                .signingRegion(resolveRegion(request))
+                                                .build();
+
+        SdkHttpFullRequest fullRequest = Aws4Signer.create().presign(httpRequest, presignRequest);
+        String signedUrl = fullRequest.getUri().toString();
+
+        // Format should be: <hostname>>:<port>>/?Action=connect&DBUser=<username>>&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expi...
+        // Note: This must be the real RDS hostname, not proxy or tunnels
+        return StringUtils.replacePrefixIgnoreCase(signedUrl, "https://", "");
+    }
+
+    private Region resolveRegion(GenerateAuthenticationTokenRequest request) {
+        if (request.region() != null) {
+            return request.region();
+        }
+
+        if (this.clientConfiguration != null && this.clientConfiguration.option(AwsClientOption.AWS_REGION) != null) {
+            return this.clientConfiguration.option(AwsClientOption.AWS_REGION);
+        }
+
+        throw new IllegalArgumentException("Region should be provided either in GenerateAuthenticationTokenRequest object or "
+                                           + "inherited from RdsClient#utilities");
+    }
+
+    private AwsCredentialsProvider resolveCredentials(GenerateAuthenticationTokenRequest request) {
+        if (request.credentialsProvider() != null) {
+            return request.credentialsProvider();
+        }
+
+        if (this.clientConfiguration != null && this.clientConfiguration.option(AwsClientOption.CREDENTIALS_PROVIDER) != null) {
+            return this.clientConfiguration.option(AwsClientOption.CREDENTIALS_PROVIDER);
+        }
+
+        throw new IllegalArgumentException("CredentialProvider should be provided either in GenerateAuthenticationTokenRequest "
+                                           + "object or inherited from RdsClient#utilities");
+    }
+
+    /**
+     * Builder class to construct {@link RdsUtilities} object
+     */
+    public static final class Builder {
+        private SdkClientConfiguration clientConfiguration;
+
+        private Builder() {
+        }
+
+        private Builder clientConfiguration(SdkClientConfiguration clientConfiguration) {
```

Comment:
`SdkClientConfiguration` is a protected API, so not sure if we want to use it here
It seems we only need region and credential provider here

Reply:
Was worried about future compatibility, but since this is private we can worry about that if/when that time comes

Reply:
Builder now stores the cred provider and region as fields. This method is used by the package protected method `RdsUtilities#create` which is generated to be used from `RdsClient#utiltiies()`

DefaultRdsClient.java
```
    @Override
    public RdsUtilities utilities() {
        return RdsUtilities.create(clientConfiguration);
    }
```

So everything is now hidden which should be hidden.